.travis.yml

```diff
@@ -80,30 +90,42 @@ install:
   - bash CI/travis/install-ptf.sh
   - bash CI/travis/install-bmv2.sh
   - bash CI/travis/install-p4c-bmv2.sh
+  - bash CI/travis/install-p4ofagent.sh
+  - export CPPFLAGS="$CPPFLAGS -I$HOME/p4ofagent/include"
+  - export LDFLAGS="$LDFLAGS -L$HOME/p4ofagent/lib"
+  - git clone https://github.com/floodlight/oftest.git oftest_tmp
   - export PYTHONPATH=$PYP_ADD:$PYTHONPATH
   - sudo ldconfig
-  - ./autogen.sh; ./configure --with-bmv2 --with-switchsai; make -j2
+  - ./autogen.sh
 
 before_script:
   - cp CI/travis/veth_setup.sh tools/
   - cd tools; sudo ./veth_setup.sh; sudo ./veth_disable_ipv6.sh; cd ..
 
-# TODO: improve
 script:
+  - $SAI_BUILD
+  - $OF_BUILD
+  - make -j2
+  - sudo ./bmv2/run_bm.sh &>/dev/null &
+  - sleep 10
+  - sudo ./bmv2/run_drivers.sh &>/dev/null &
+  - sleep 10
+  - $SAI_TEST_SCRIPT
+  - sudo killall bmswitchp4_drivers lt-bmswitchp4_drivers simple_switch lt-simple_switch; echo 0
   - sudo ./bmv2/run_bm.sh &>/dev/null &
   - sleep 10
   - sudo ./bmv2/run_drivers.sh &>/dev/null &
   - sleep 10
-  - sudo PYTHONPATH=$PYP_ADD:$PYTHONPATH ./bmv2/run_tests.sh
+  - $API_TEST_SCRIPT
   - sudo killall bmswitchp4_drivers lt-bmswitchp4_drivers simple_switch lt-simple_switch; echo 0
   - sudo ./bmv2/run_bm.sh &>/dev/null &
   - sleep 10
   - sudo ./bmv2/run_drivers.sh &>/dev/null &
   - sleep 10
-  - sudo PYTHONPATH=$PYP_ADD:$PYTHONPATH ./bmv2/run_tests.sh --test-dir tests/ptf-tests/api-tests/
+  - $PD_TEST_SCRIPT
   - sudo killall bmswitchp4_drivers lt-bmswitchp4_drivers simple_switch lt-simple_switch; echo 0
   - sudo ./bmv2/run_bm.sh &>/dev/null &
   - sleep 10
   - sudo ./bmv2/run_drivers.sh &>/dev/null &
   - sleep 10
-  - sudo PYTHONPATH=$PYP_ADD:$PYTHONPATH ./bmv2/run_tests.sh --test-dir tests/ptf-tests/sai-tests
+  - $OF_TEST_SCRIPT
```

Comment:
is there a reason why you did not move the testing code for PD+API+SAI and the testing code for OF to separate scripts under `CI/travis`, e.g. `run_switch_test.sh` and `run_of_test.sh`. Then this whole Travis file becomes much simpler.
Something like this:
```
matrix:
  - TEST_SCENARIO="switch"
  - TEST_SCENARIO="of"
...
script:
  - if [ "$TEST_SCENARIO" != "switch" ]; then ./configure --with-bmv2 --with-switchsai; fi
  - if [ "$TEST_SCENARIO" != "of" ]; then ./configure --with-bmv2 --with-of; fi
  - if [ "$TEST_SCENARIO" != "switch" ]; then ./CI/travis/run_switch_test.sh; fi
  - if [ "$TEST_SCENARIO" != "of" ]; then ./CI/travis/run_of_test.sh; fi
```
It seems that with the current file, you sometimes start the switch & drivers (+ sleep) for nothing.